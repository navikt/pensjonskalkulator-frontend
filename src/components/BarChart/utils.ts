import { ChartData } from './BarChart'

export const getBarChartHeight = (
  value: number,
  maxValue: number,
  maxHeight: number = 128
) => {
  if (value === 0) {
    return 0
  }
  return (value / maxValue) * maxHeight
}

export const findMaxValue = (data: ChartData[]): number => {
  return data.reduce((max, { value }) => (max > value ? max : value), 0)
}
