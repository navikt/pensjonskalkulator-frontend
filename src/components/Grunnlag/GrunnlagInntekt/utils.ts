export const validateInntektInput = (
  s: string | undefined,
  updateValidationErrorMessage: (s: string) => void
) => {
  let isValid = true
  if (s === undefined || s === '') {
    isValid = false
    updateValidationErrorMessage(
      'grunnlag.inntekt.inntektmodal.textfield.validation_error.required'
    )
  } else if (isNaN(s as unknown as number) || !/^[0-9]+$/.test(s)) {
    isValid = false
    updateValidationErrorMessage(
      'grunnlag.inntekt.inntektmodal.textfield.validation_error.type'
    )
  } else if (parseInt(s as string, 10) > 100000000) {
    isValid = false
    updateValidationErrorMessage(
      'grunnlag.inntekt.inntektmodal.textfield.validation_error.max'
    )
  }
  return isValid
}
