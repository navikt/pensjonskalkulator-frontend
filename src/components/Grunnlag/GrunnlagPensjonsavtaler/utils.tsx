import { PensjonsavtaleKategori } from '@/types/enums'

type PensjonsavtaleRecord = {
  [key in PensjonsavtaleKategori]?: Pensjonsavtale[]
}

export const groupPensjonsavtalerByType = (
  pensjonsavtaler: Pensjonsavtale[]
): PensjonsavtaleRecord => {
  const record: PensjonsavtaleRecord = {}

  for (const avtale of pensjonsavtaler) {
    const avtalekategori =
      PensjonsavtaleKategori[
        avtale.kategori as keyof typeof PensjonsavtaleKategori
      ]

    if (!Array.isArray(record[avtalekategori])) {
      record[avtalekategori] = [avtale]
    } else {
      record[avtalekategori]?.push(avtale)
    }
  }
  return record
}

export function getMaanedString(maaned?: number) {
  if (maaned !== undefined && maaned > 0) {
    return ` og ${maaned} md.`
  }
  return ''
}
