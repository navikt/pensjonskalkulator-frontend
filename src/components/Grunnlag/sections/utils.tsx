import { PensjonsavtaleKategori } from '@/types/enums'

type PensjonsavtaleRecord = {
  [key in PensjonsavtaleKategori]?: Pensjonsavtale[]
}

export const groupPensjonsavtalerByType = (
  pensjonsavtaler: Pensjonsavtale[]
): PensjonsavtaleRecord => {
  const record: PensjonsavtaleRecord = {}

  for (const avtale of pensjonsavtaler) {
    const avtalekategori =
      PensjonsavtaleKategori[
        avtale.kategori as keyof typeof PensjonsavtaleKategori
      ]

    if (!Array.isArray(record[avtalekategori])) {
      record[avtalekategori] = [avtale]
    } else {
      record[avtalekategori]?.push(avtale)
    }
  }
  return record
}

export function getPensjonsavtalerTittel(
  harSamtykket: boolean,
  showError: boolean,
  antallAvtaler: string
) {
  if (!harSamtykket) {
    return 'Ikke innhentet'
  }
  if (showError) {
    return 'Kan ikke hentes'
  }
  return antallAvtaler
}
