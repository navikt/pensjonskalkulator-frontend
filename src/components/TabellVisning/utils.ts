import { IntlShape } from 'react-intl'

import { SeriesColumnOptions } from 'highcharts'

export type TableDataRowDetaljer = { name: string; subSum: number }

export type TableDataRow = {
  alder: string
  sum: number
  detaljer: TableDataRowDetaljer[]
}

export function formatSeriesToTableData(
  intl: IntlShape,
  series: SeriesColumnOptions[],
  aarArray?: string[]
): TableDataRow[] {
  const tableData: TableDataRow[] = []
  if (aarArray === undefined) {
    return tableData
  }

  for (let i = 0; i < aarArray.length; i++) {
    let sum = 0
    const detaljer: TableDataRowDetaljer[] = []
    const alder = `${aarArray[i]} ${
      aarArray[i].includes('+')
        ? intl.formatMessage({ id: 'alder.aar_livsvarig' })
        : intl.formatMessage({ id: 'alder.aar' })
    }`

    for (const obj of series) {
      const { name, data } = obj
      if (data && name) {
        const temporarySum = data[i] as number
        sum += temporarySum

        detaljer.push({ name, subSum: temporarySum })
      }
    }

    tableData.push({ alder, sum, detaljer })
  }

  return tableData
}
