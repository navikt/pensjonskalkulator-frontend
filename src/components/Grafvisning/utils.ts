import React from 'react'

import {
  Axis,
  AxisLabelsFormatterContextObject,
  Options,
  Point,
  SeriesColumnOptions,
  TooltipFormatterContextObject,
} from 'highcharts'

import { formatAsDecimal } from '@/utils/currency'

import globalClassNames from '*.module.scss'

export const highchartsScrollingSelector = '.highcharts-scrolling'
export const COLUMN_WIDTH = 25
const TOOLTIP_YPOS = 35

export function labelFormatter(this: AxisLabelsFormatterContextObject) {
  return this.value > 1000
    ? ((this.value as number) / 1000).toString()
    : this.value.toString()
}

export type ExtendedYAxis = Axis & { height: number; pos: number }
export type ExtendedPoint = Point & { tooltipPos: number[] }

export function tooltipFormatter(
  context: TooltipFormatterContextObject,
  styles: Partial<typeof globalClassNames>
): string {
  const yAxisHeight = (context.points?.[0].series.yAxis as ExtendedYAxis).height
  const lineYpos =
    (context.points?.[0].series.chart.yAxis[0] as ExtendedYAxis).pos -
    TOOLTIP_YPOS
  const columnHeight =
    yAxisHeight - (context.points?.[0].point as ExtendedPoint).tooltipPos[1]
  const scrollPosition =
    document.querySelector(highchartsScrollingSelector)?.scrollLeft ?? 0

  const leftPosition = context.points?.[0].point?.plotX ?? 0
  const tooltipConnectingLine = `<div class="${
    styles.tooltipLine
  }" style="top: ${lineYpos}px; left: ${
    leftPosition + 21 - scrollPosition + COLUMN_WIDTH / 2
  }px; height: ${yAxisHeight - columnHeight}px"></div>`

  const headerFormat =
    `<table class="${styles.tooltipTable}"><thead><tr>` +
    `<th class="${styles.tooltipTableHeaderCell} ${styles.tooltipTableHeaderCell__left}">Pensjon og inntekt det året du er ${context.x} år</th>` +
    `<th class="${styles.tooltipTableHeaderCell} ${
      styles.tooltipTableHeaderCell__right
    }">${formatAsDecimal(context.points?.[0].total)} kr</th>` +
    `</tr></thead><tbody>`

  let pointsFormat = ''
  context?.points?.forEach(function (point) {
    pointsFormat +=
      `<tr>` +
      `<td class="${styles.tooltipTableCell}"><span class="${styles.tooltipTableCellDot}" style="background-color:${point.series.color}"></span>${point.series.name}</td>` +
      `<td class="${styles.tooltipTableCell} ${
        styles.tooltipTableCell__right
      }">${formatAsDecimal(point.y)} kr</td>` +
      `</tr>`
  })

  const footerFormat = '</tbody></table>'
  return `${headerFormat}${pointsFormat}${footerFormat}${tooltipConnectingLine}`
}

export const onVisFaerreAarClick = () => {
  const el = document.querySelector(highchartsScrollingSelector)
  if (el) {
    ;(el as HTMLElement).scrollLeft -= 50
  }
}

export const onVisFlereAarClick = () => {
  const el = document.querySelector(highchartsScrollingSelector)
  if (el) {
    ;(el as HTMLElement).scrollLeft += 50
  }
}

type HighchartsScrollingHTMLDivElement = HTMLDivElement & {
  handleButtonVisibility: {
    showRightButton: React.Dispatch<React.SetStateAction<boolean>>
    showLeftButton: React.Dispatch<React.SetStateAction<boolean>>
  }
}

export function handleChartScroll(event: Event) {
  if (event.currentTarget) {
    const el = event.currentTarget as HighchartsScrollingHTMLDivElement
    const elementScrollPosition = el.scrollLeft

    if (elementScrollPosition === 0) {
      el.handleButtonVisibility.showRightButton(true)
      el.handleButtonVisibility.showLeftButton(false)
    } else if (elementScrollPosition + el.offsetWidth === el.scrollWidth) {
      el.handleButtonVisibility.showRightButton(false)
      el.handleButtonVisibility.showLeftButton(true)
    } else {
      el.handleButtonVisibility.showRightButton(true)
      el.handleButtonVisibility.showLeftButton(true)
    }
  }
}

export const getChartOptions = (
  styles: Partial<typeof globalClassNames>,
  showRightButton: React.Dispatch<React.SetStateAction<boolean>>,
  showLeftButton: React.Dispatch<React.SetStateAction<boolean>>
): Options => {
  return {
    chart: {
      type: 'column',
      spacingTop: 0,
      spacingLeft: 0,
      spacingRight: 25,
      scrollablePlotArea: {
        minWidth: 750,
        scrollPositionX: 0,
      },
      events: {
        render() {
          const highchartsScrollingElement = document.querySelector(
            highchartsScrollingSelector
          )

          if (highchartsScrollingElement) {
            const el =
              highchartsScrollingElement as HighchartsScrollingHTMLDivElement
            // Denne setTimeout er nødvendig fordi highcharts tegner scroll container litt etter render callback og har ikke noe eget flag for den
            setTimeout(() => {
              const elementScrollWidth = el.scrollWidth
              const elementWidth = el.offsetWidth
              showRightButton(elementScrollWidth > elementWidth)

              el.addEventListener('scroll', handleChartScroll, false)
              el.handleButtonVisibility = {
                showRightButton,
                showLeftButton,
              }
            }, 0)
          } else {
            showRightButton(false)
            showLeftButton(false)
          }
        },
      },
    },
    title: {
      text: `Beregning`,
      align: 'left',
      margin: 40,
      y: 20,
      style: {
        fontFamily: 'var(--a-font-family)',
        fontWeight: 'bold',
        fontSize: '20px',
      },
    },
    xAxis: {
      categories: [],
    },
    yAxis: {
      gridLineDashStyle: 'Dash',
      minorTickInterval: 200000,
      tickInterval: 200000,
      allowDecimals: false,
      min: 0,
      title: {
        text: 'Tusen kroner',
        align: 'high',
        offset: -55,
        rotation: 0,
        x: -12,
        y: -20,
      },
      labels: {
        formatter: labelFormatter,
      },
    },
    credits: {
      enabled: false,
    },
    tooltip: {
      useHTML: true,
      className: styles.tooltip,
      /* c8 ignore next 3 */
      formatter: function (this: TooltipFormatterContextObject) {
        return tooltipFormatter(this, styles)
      },
      outside: true,
      shadow: false,
      shared: true,
      padding: 0,
      /* c8 ignore next 3 */
      positioner: function () {
        return { x: 0, y: TOOLTIP_YPOS }
      },
    },
    legend: {
      useHTML: true,
      x: 0,
      y: -25,
      padding: 0,
      margin: 50,
      layout: 'horizontal',
      align: 'left',
      verticalAlign: 'top',
      itemDistance: 0,
      itemStyle: {
        fontFamily: 'var(--a-font-family)',
        color: '#000000',
        fontWeight: 'regular',
        fontSize: '14px',
        cursor: 'default',
        zIndex: 0,
      },
      itemHoverStyle: { color: '#000000' },
      itemMarginBottom: 5,
    },
    plotOptions: {
      series: {
        stacking: 'normal',
        states: {
          inactive: {
            enabled: false,
          },
        },
        events: {
          /* c8 ignore next 3 */
          legendItemClick: function (e) {
            e.preventDefault()
          },
        },
      },
    },
    series: [],
  }
}

export const removeHandleChartScrollEventListener = () => {
  const el = document.querySelector(highchartsScrollingSelector)
  el?.removeEventListener('scroll', handleChartScroll)
}

export const getAlderspensjonColumnOptions = (
  data: SeriesColumnOptions['data']
): SeriesColumnOptions => ({
  type: 'column',
  pointWidth: COLUMN_WIDTH,
  name: 'Alderspensjon (NAV)',
  color: 'var(--a-deepblue-500)',
  states: {
    hover: {
      color: 'var(--a-deepblue-200)',
    },
  },
  data,
})

export const getAfpSeriesData = (
  data: SeriesColumnOptions['data']
): SeriesColumnOptions => ({
  type: 'column',
  pointWidth: COLUMN_WIDTH,
  name: 'Avtalefestet pensjon (AFP)',
  color: 'var(--a-purple-400)',
  states: {
    hover: {
      color: 'var(--a-purple-200)',
    },
  },
  data,
})

export const getInntektSeriesData = (
  data: SeriesColumnOptions['data']
): SeriesColumnOptions => ({
  type: 'column',
  pointWidth: COLUMN_WIDTH,
  name: 'Inntekt (lønn m.m.)',
  color: '#868F9C',
  data,
})

export const getPensjonsavtalerSeriesData = (
  data: SeriesColumnOptions['data']
): SeriesColumnOptions => ({
  type: 'column',
  pointWidth: COLUMN_WIDTH,
  name: 'Pensjonsavtaler (arbeidsgiver)',
  color: 'var(--a-green-400)',
  states: {
    hover: {
      color: 'var(--a-green-200)',
    },
  },
  data,
})

export const hasSeriesData = (data: SeriesColumnOptions['data']) => {
  return data?.some((it) => typeof it === 'number' && it > 0)
}

export const isSeriesColumnOptions = (
  options: any
): options is SeriesColumnOptions => {
  return (
    options?.type === 'column' &&
    Array.isArray(options.data) &&
    options.data.length > 0
  )
}
