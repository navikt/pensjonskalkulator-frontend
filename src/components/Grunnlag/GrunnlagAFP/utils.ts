// Input:
// AFPValg
// isKap19
// erApoteker
// loependeVedtak
// uforegrad
// intl - interasjonalisering
import { IntlShape } from 'react-intl'

import { isFoedtFoer1963 } from '@/utils/alder'

interface IAfpGrunnlagInput {
  /** Dersom bruker er kap19/apoteker kan man velge å beregne offentlig AFP sammen med aldersensjon */
  afpUtregning: AfpUtregningValg
  /** Valg av hvilken AFP utregning bruker har valgt */
  afpValg: AfpRadio | null
  /** Valg for beregning av pensjon */
  beregningsvalg: Beregningsvalg | null
  /** Indikerer om brukeren er apoteker */
  erApoteker: boolean
  /** Brukerens fødselsdato */
  foedselsdato: string
  /** Informasjon om løpende vedtak */
  loependeVedtak: LoependeVedtak
  /** Samtykke til beregning av offentlig AFP */
  samtykkeOffentligAFP: boolean | null
}

export const afpContentIntl = (intl: IntlShape) => ({
  offentligIkkeBeregnet_1: {
    title: `${intl.formatMessage({ id: 'afp.offentlig' })} (${intl.formatMessage({ id: 'grunnlag.afp.ikke_beregnet' })})`,
    content: 'grunnlag.afp.ingress.ja_offentlig_utilgjengelig',
  },

  afpOffentligOppgitt_2: {
    title: `${intl.formatMessage({ id: 'afp.offentlig' })}`,
    content: 'grunnlag.afp.ingress.ja_offentlig',
  },

  offentligAfpOgUforeKanIkkeBeregnes_3: {
    title: `${intl.formatMessage({ id: 'afp.offentlig' })} (${intl.formatMessage({ id: 'grunnlag.afp.ikke_beregnet' })})`,
    content: 'grunnlag.afp.ingress.ja_offentlig.ufoeretrygd',
  },

  afpPrivat_4: {
    title: `${intl.formatMessage({ id: 'afp.privat' })}`,
    content: 'grunnlag.afp.ingress.ja_privat',
  },

  afpPrivatIkkeBeregnet_5: {
    title: `${intl.formatMessage({ id: 'afp.nei' })} (${intl.formatMessage({ id: 'grunnlag.afp.ikke_beregnet' })})`,
    content: 'grunnlag.afp.ingress.ja_privat.ufoeretrygd',
  },

  afpIkkeSvart_6: {
    title: `${intl.formatMessage({ id: 'afp.nei' })}`,
    content: 'grunnlag.afp.ingress.nei',
  },

  afpVetIkke_7: {
    title: `${intl.formatMessage({ id: 'afp.vet_ikke' })}`,
    content: 'grunnlag.afp.ingress.vet_ikke',
  },

  afpVetIkkeUforetrygd_8: {
    title: `${intl.formatMessage({ id: 'afp.vet_ikke' })}`,
    content: 'grunnlag.afp.ingress.vet_ikke.ufoeretrygd',
  },

  afpUforetrygd_9: {
    title: `${intl.formatMessage({ id: 'afp.nei' })}`,
    content: 'grunnlag.afp.ingress.overgangskull.ufoeretrygd_eller_ap',
  },

  afpUforetrygdNei_10: {
    title: `${intl.formatMessage({ id: 'afp.nei' })}`,
    content: 'grunnlag.afp.ingress.ufoeretrygd',
  },
  afpPrivatUendret_11: {
    title: `${intl.formatMessage({ id: 'afp.privat' })} (${intl.formatMessage({ id: 'grunnlag.afp.endring' })})`,
    content: 'grunnlag.afp.ingress.ja_privat.endring',
  },
  harAfpOffentlig_12: {
    title: `${intl.formatMessage({ id: 'afp.offentlig' })}`,
    content: 'grunnlag.afp.ingress.overgangskull',
  },
})

export const generateAfpContent =
  (intl: IntlShape) => (input: IAfpGrunnlagInput) => {
    const content = afpContentIntl(intl)
    const {
      afpValg,
      samtykkeOffentligAFP,
      afpUtregning,
      foedselsdato,
      loependeVedtak,
      beregningsvalg,
      erApoteker,
    } = input

    const hasUfoeregradGreaterThanZero = loependeVedtak?.ufoeretrygd?.grad > 0
    const has100PercentUfoeretrygd = loependeVedtak?.ufoeretrygd?.grad === 100
    const hasLoependeVedtak = loependeVedtak?.harLoependeVedtak === true
    const isKap19 = isFoedtFoer1963(foedselsdato)
    const hasAfpPrivat = loependeVedtak?.afpPrivat !== undefined
    const hasAfpOffentlig = loependeVedtak?.afpOffentlig !== undefined
    const hasAlderspensjon = loependeVedtak?.alderspensjon !== undefined
    const isAfpValgMissing = afpValg === null || afpValg === undefined

    // Personer født før 1963 eller apotekere behandles likt for AFP-formål
    const isKap19OrApoteker = isKap19 || erApoteker

    // Prioritet 1: Håndter eksisterende AFP-vedtak
    if (hasAfpOffentlig) {
      return content.harAfpOffentlig_12
    }
    if (hasAfpPrivat) {
      return content.afpPrivatUendret_11
    }

    // Prioritet 2: Håndter spesielle regler for apotekere og personer født før 1963
    if (isKap19OrApoteker) {
      // Offentlig AFP med KUN_ALDERSPENSJON
      if (afpValg === 'ja_offentlig' && afpUtregning === 'KUN_ALDERSPENSJON') {
        return content.afpIkkeSvart_6
      }

      // Manglende afpValg med vedtak om alderspensjon eller uføretrygd
      if (
        afpValg === null &&
        (hasAlderspensjon || hasUfoeregradGreaterThanZero)
      ) {
        return content.afpUforetrygd_9
      }
    }

    // Prioritet 3: Håndter tilfeller med uføretrygd
    if (hasUfoeregradGreaterThanZero) {
      // Tilfelle 3.1: Født før 1963 eller apoteker med manglende AFP-valg
      if (isKap19OrApoteker && isAfpValgMissing) {
        return content.afpUforetrygd_9
      }

      // Tilfelle 3.2: Personer som verken er født før 1963 eller er apoteker, med uføretrygd
      if (!isKap19OrApoteker) {
        // Håndter ulike AFP-valg med uføretrygd
        switch (afpValg) {
          case 'ja_offentlig': {
            // Manglende samtykke til AFP-beregning (prioriteres høyest)
            if (samtykkeOffentligAFP === false) {
              return content.offentligIkkeBeregnet_1
            }

            if (beregningsvalg === 'uten_afp' || beregningsvalg === null) {
              return content.offentligAfpOgUforeKanIkkeBeregnes_3
            }

            return content.afpOffentligOppgitt_2
          }

          case 'ja_privat': {
            if (beregningsvalg === 'uten_afp') {
              return content.afpPrivatIkkeBeregnet_5
            }
            return content.afpPrivat_4
          }

          case 'nei': {
            // Spesialtilfelle for gradert vedtak uten 100% uføregrad
            if (!has100PercentUfoeretrygd && !isKap19) {
              return content.afpIkkeSvart_6
            }
            return content.afpUforetrygdNei_10
          }

          case 'vet_ikke': {
            return content.afpVetIkkeUforetrygd_8
          }

          default: {
            // Tilfelle for null/undefined
            if (!isKap19OrApoteker && (hasLoependeVedtak || hasAlderspensjon)) {
              return content.afpUforetrygdNei_10
            }
            return content.afpUforetrygdNei_10
          }
        }
      }
    }

    // Prioritet 4: Håndter AFP-valg basert på forskjellige betingelser (for tilfeller uten uføretrygd)
    switch (afpValg) {
      case 'ja_offentlig': {
        // Tilfelle 4.1: Født 1963 eller senere (og ikke apoteker) med offentlig AFP
        if (!isKap19OrApoteker) {
          // Manglende samtykke til AFP-beregning
          if (samtykkeOffentligAFP === false) {
            return content.offentligIkkeBeregnet_1
          }

          // Samtykket til AFP-beregning
          return content.afpOffentligOppgitt_2
        }

        // Tilfelle 4.2: Født før 1963 eller apoteker med offentlig AFP (isKap19OrApoteker)
        if (afpUtregning === 'KUN_ALDERSPENSJON') {
          return content.afpIkkeSvart_6
        }

        if (hasLoependeVedtak && hasUfoeregradGreaterThanZero) {
          return content.harAfpOffentlig_12
        }

        return content.afpOffentligOppgitt_2
      }

      case 'ja_privat': {
        // Tilfelle 4.3: Privat AFP (standardrespons)
        return content.afpPrivat_4
      }

      case 'nei': {
        // Tilfelle 4.4: Ingen AFP
        return content.afpIkkeSvart_6
      }

      case 'vet_ikke':
      default: {
        // Tilfelle 4.5: "Vet ikke"
        return content.afpVetIkke_7
      }
    }
  }
