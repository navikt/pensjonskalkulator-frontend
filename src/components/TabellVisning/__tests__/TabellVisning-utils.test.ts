import { SeriesColumnOptions } from 'highcharts'
import { describe, expect, it } from 'vitest'

import { formatSeriesToTableData } from '../utils'

describe('TabellVisning-utils', () => {
  describe('formatSeriesToTableData', () => {
    const alderArray = ['62', '63', '64', '65', '66', '67', '68', '68', '70+']
    const series: SeriesColumnOptions[] = [
      {
        type: 'column',
        name: 'Pensjonsgivende inntekt',
        data: [
          100000, 175000, 220000, 220000, 220000, 220000, 220000, 220000,
          220000,
        ],
      },
      {
        type: 'column',
        name: 'AFP (Avtalefestet pensjon)',
        data: [
          150000, 200000, 350000, 350000, 350000, 350000, 350000, 350000,
          350000,
        ],
      },
      {
        type: 'column',
        name: 'Pensjonsavtaler (arbeidsgivere m.m.)',
        data: [
          180000, 250000, 380000, 380000, 380000, 380000, 380000, 380000,
          380000,
        ],
      },
      {
        type: 'column',
        name: 'Alderspensjon (NAV)',
        data: [
          200000, 350000, 400000, 400000, 400000, 400000, 400000, 400000,
          400000,
        ],
      },
    ]
    it('returnerer tomt array når alder array er undefined eller tomt', () => {
      const result0 = formatSeriesToTableData([])
      expect(result0).toHaveLength(0)
      const result1 = formatSeriesToTableData([], [])
      expect(result1).toHaveLength(0)
      const result2 = formatSeriesToTableData([...series], [])
      expect(result2).toHaveLength(0)
    })

    it('returnerer et array med riktig struktur og headere når alder array ikke er tomt', () => {
      const result = formatSeriesToTableData([], [...alderArray])
      expect(result).toHaveLength(alderArray.length)
      expect(result[0].alder).toBe('62 år')
      expect(result[alderArray.length - 1].alder).toBe('70+ år (livsvarig)')
      expect(result).toMatchSnapshot()
    })

    it('returnerer et array med riktig struktur og detaljer når begge arrayeme haer data', () => {
      const result = formatSeriesToTableData([...series], [...alderArray])
      expect(result).toHaveLength(alderArray.length)
      expect(result[0].detaljer).toHaveLength(series.length)
      expect(result).toMatchSnapshot()
    })
  })
})
