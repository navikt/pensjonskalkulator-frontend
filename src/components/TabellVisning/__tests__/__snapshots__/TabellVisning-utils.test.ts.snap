// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TabellVisning-utils > formatSeriesToTableData > returnerer et array med riktig struktur og detaljer når begge arrayeme haer data 1`] = `
[
  {
    "alder": "62 år",
    "detaljer": [
      {
        "name": "Pensjonsgivende inntekt",
        "subSum": 100000,
      },
      {
        "name": "AFP (Avtalefestet pensjon)",
        "subSum": 150000,
      },
      {
        "name": "Pensjonsavtaler (arbeidsgivere m.m.)",
        "subSum": 180000,
      },
      {
        "name": "Alderspensjon (NAV)",
        "subSum": 200000,
      },
    ],
    "sum": 630000,
  },
  {
    "alder": "63 år",
    "detaljer": [
      {
        "name": "Pensjonsgivende inntekt",
        "subSum": 175000,
      },
      {
        "name": "AFP (Avtalefestet pensjon)",
        "subSum": 200000,
      },
      {
        "name": "Pensjonsavtaler (arbeidsgivere m.m.)",
        "subSum": 250000,
      },
      {
        "name": "Alderspensjon (NAV)",
        "subSum": 350000,
      },
    ],
    "sum": 975000,
  },
  {
    "alder": "64 år",
    "detaljer": [
      {
        "name": "Pensjonsgivende inntekt",
        "subSum": 220000,
      },
      {
        "name": "AFP (Avtalefestet pensjon)",
        "subSum": 350000,
      },
      {
        "name": "Pensjonsavtaler (arbeidsgivere m.m.)",
        "subSum": 380000,
      },
      {
        "name": "Alderspensjon (NAV)",
        "subSum": 400000,
      },
    ],
    "sum": 1350000,
  },
  {
    "alder": "65 år",
    "detaljer": [
      {
        "name": "Pensjonsgivende inntekt",
        "subSum": 220000,
      },
      {
        "name": "AFP (Avtalefestet pensjon)",
        "subSum": 350000,
      },
      {
        "name": "Pensjonsavtaler (arbeidsgivere m.m.)",
        "subSum": 380000,
      },
      {
        "name": "Alderspensjon (NAV)",
        "subSum": 400000,
      },
    ],
    "sum": 1350000,
  },
  {
    "alder": "66 år",
    "detaljer": [
      {
        "name": "Pensjonsgivende inntekt",
        "subSum": 220000,
      },
      {
        "name": "AFP (Avtalefestet pensjon)",
        "subSum": 350000,
      },
      {
        "name": "Pensjonsavtaler (arbeidsgivere m.m.)",
        "subSum": 380000,
      },
      {
        "name": "Alderspensjon (NAV)",
        "subSum": 400000,
      },
    ],
    "sum": 1350000,
  },
  {
    "alder": "67 år",
    "detaljer": [
      {
        "name": "Pensjonsgivende inntekt",
        "subSum": 220000,
      },
      {
        "name": "AFP (Avtalefestet pensjon)",
        "subSum": 350000,
      },
      {
        "name": "Pensjonsavtaler (arbeidsgivere m.m.)",
        "subSum": 380000,
      },
      {
        "name": "Alderspensjon (NAV)",
        "subSum": 400000,
      },
    ],
    "sum": 1350000,
  },
  {
    "alder": "68 år",
    "detaljer": [
      {
        "name": "Pensjonsgivende inntekt",
        "subSum": 220000,
      },
      {
        "name": "AFP (Avtalefestet pensjon)",
        "subSum": 350000,
      },
      {
        "name": "Pensjonsavtaler (arbeidsgivere m.m.)",
        "subSum": 380000,
      },
      {
        "name": "Alderspensjon (NAV)",
        "subSum": 400000,
      },
    ],
    "sum": 1350000,
  },
  {
    "alder": "68 år",
    "detaljer": [
      {
        "name": "Pensjonsgivende inntekt",
        "subSum": 220000,
      },
      {
        "name": "AFP (Avtalefestet pensjon)",
        "subSum": 350000,
      },
      {
        "name": "Pensjonsavtaler (arbeidsgivere m.m.)",
        "subSum": 380000,
      },
      {
        "name": "Alderspensjon (NAV)",
        "subSum": 400000,
      },
    ],
    "sum": 1350000,
  },
  {
    "alder": "70+ år (livsvarig)",
    "detaljer": [
      {
        "name": "Pensjonsgivende inntekt",
        "subSum": 220000,
      },
      {
        "name": "AFP (Avtalefestet pensjon)",
        "subSum": 350000,
      },
      {
        "name": "Pensjonsavtaler (arbeidsgivere m.m.)",
        "subSum": 380000,
      },
      {
        "name": "Alderspensjon (NAV)",
        "subSum": 400000,
      },
    ],
    "sum": 1350000,
  },
]
`;

exports[`TabellVisning-utils > formatSeriesToTableData > returnerer et array med riktig struktur og headere når alder array ikke er tomt 1`] = `
[
  {
    "alder": "62 år",
    "detaljer": [],
    "sum": 0,
  },
  {
    "alder": "63 år",
    "detaljer": [],
    "sum": 0,
  },
  {
    "alder": "64 år",
    "detaljer": [],
    "sum": 0,
  },
  {
    "alder": "65 år",
    "detaljer": [],
    "sum": 0,
  },
  {
    "alder": "66 år",
    "detaljer": [],
    "sum": 0,
  },
  {
    "alder": "67 år",
    "detaljer": [],
    "sum": 0,
  },
  {
    "alder": "68 år",
    "detaljer": [],
    "sum": 0,
  },
  {
    "alder": "68 år",
    "detaljer": [],
    "sum": 0,
  },
  {
    "alder": "70+ år (livsvarig)",
    "detaljer": [],
    "sum": 0,
  },
]
`;
